src/awkward/operations/ak_all.py

```diff
@@ -3,15 +3,27 @@
 import awkward as ak
 from awkward._behavior import behavior_of
 from awkward._connect.numpy import UNSUPPORTED
-from awkward._errors import with_operation_context
+from awkward._dispatch import high_level_function
 from awkward._layout import wrap_layout
 from awkward._nplikes.numpylike import NumpyMetadata
 from awkward._regularize import regularize_axis
 
 np = NumpyMetadata.instance()
 
 
-@with_operation_context
+def _dispatcher(
+    array,
+    axis=None,
+    *,
+    keepdims=False,
+    mask_identity=False,
+    highlevel=True,
+    behavior=None,
+):
+    yield array
+
+
+@high_level_function(_dispatcher)
 def all(
     array,
     axis=None,
```

Comment:
By adding `with_operation_context`, you managed to eliminate the need to manually duplicate each high-level signature, but it looks like `_dispatcher` make you have to put them all back. What's worse, now you also need to specify the defaults twice.

If you can extract the signature from third-party overloaded `__awkward_function__`s, can't you extract the signature from our own high-level functions?

You can chain decorators, if that helps: it doesn't need to be passed as an argument to `high_level_function`. Would there be a way to do it like

```python
@with_operation_context
@_dispatcher
def all(
    ...
)
```

and have `_dispatcher` pull the signature off of the `all` function that gets passed to it?

Reply:
The purpose of `dispatcher` is to make Awkward's mechanism aware of which arguments are array-like. We don't need to use an identical signature, but if the dispatch mechanism is a separate function, then any modification to the signature (such as permitting the dispatcher to just use non-keyword-only, non-position-only arguments arguments) requires converting between the two. This would be done using `inspect.Signature.bind()`, which adds some overhead (I mention this [here](https://github.com/scikit-hep/awkward/pull/2530#issue-1763535738)). I'm not yet clear on where the equilibrium point is between convenience & expressiveness vs performance in this aspect of the code, but being mindful of the graph optimisation cost, I am reluctant to make things slower out of the gate.

We can otherwise use a naive implementation that is permitted to look for arrays in tuples and lists. I don't think that's worth pursuing though; there will always be types e.g. iterators that we can't inspect, and it seems silly not to encode up-front the knowledge about which arguments (or their contents) are arrays.

Another approach that avoids the problem entirely is to just do everything inline: https://github.com/scikit-hep/awkward/pull/2531

Reply:
> it doesn't need to be passed as an argument to high_level_function. Would there be a way to do it like

Yes, originally I had this inverted; the decorator was added to `ak.any`, and then 
```python
@ak.any.dispatcher
def _dispatcher(...):
    ...
```

I ended up preferring the up-frontness of this PR (inspired by NumPy's implementation).

Reply:
Manually duplicated arguments look bad in the source code, but if they're automatically checked, then they're not actually-bad for maintenance. Let's go with this, then.